### PR TITLE
Display last message date on the list of chats

### PR DESCRIPTION
**Pull request by someone857**
Similar to Instagram native client, we will display the date when the last message was sent on the chat list. A watcher will refresh the time every couple of seconds.

### CODE REVIEW

**Comment by someone857** [comment]
@ifedapoolarewaju this is now ready to be reviewed

File: browser/js/funcs.js [hunk 1]
```diff
@@ -1,3 +1,10 @@
+const TimeAgo = require('javascript-time-ago');
```
Comment (ifedapoolarewaju) [inline browser/js/funcs.js:1]: is it fine if we use [moment.js](https://momentjs.com/docs/#/displaying/fromnow/) for this instead? It's more popular, so I'd be more comfortable with introducing that into the project

File: browser/js/index.js [hunk 2]
```diff
@@ -198,4 +211,9 @@ document.addEventListener('DOMContentLoaded', () => {
 
   getLoggedInUser();
   getChatList();
+
+  ipcRenderer.on('updatePollingInterval', (_, interval) => {
+    pollingInterval = interval;
+  });
+  setTimeout(updateDates, pollingInterval);
```
Comment (ifedapoolarewaju) [inline browser/js/index.js:218]: sorry I didn't realise this before, but I think there may be no need to poll this separately. Maybe we can just rely on the polling done by [getChatList](https://github.com/igdmapps/igdm/pull/1289/files#diff-9e4e344249ee3bd8855fbd8d4636d08fR213)? And so when the chat list updates, the time render also updates automatically.

File: browser/js/renderers.js [hunk 3]
```diff
@@ -296,15 +296,15 @@ function renderVideoContextMenu (videoUrl) {
...
+function renderChatListItem (chatTitle, msgPreview, thumbnail, id, direction, timestamp) {
   let li = document.createElement('li');
   li.classList.add('col-12', 'p-3');
 
   const msgPreviewClass = (direction == 'outward') ? 'outward' : 'inward';
 
   li.appendChild(createThumbnailDom(thumbnail));
 
-  li.appendChild(dom(`<div class="username ml-3 d-none d-sm-inline-block"><b>${chatTitle}</b><br><span class="${msgPreviewClass}">${msgPreview}</span></div>`));
+  li.appendChild(dom(`<div class="username ml-3 d-none d-sm-inline-block"><b>${chatTitle}</b><br><span class="${msgPreviewClass}">${msgPreview}</span>${renderMessageTimeAgo(timestamp)}</div>`));
```
Comment (ifedapoolarewaju) [inline browser/js/renderers.js:307]: as this line is now getting really long, maybe we should break the html snipped into multi-line?

File: browser/js/funcs.js (hunk 1 above)
Comment (someone857) [reply browser/js/funcs.js:1]: i tried it brfore but couldnt get it working for some reason. let me try again and share if i have any issues.

File: browser/js/index.js (hunk 2 above)
Comment (someone857) [reply browser/js/index.js:218]: my thought was: this is not really part of the chat and this watcher could have a different polling interval. suppose in the future you want to display dates in other places and have them watched and updated automatically. at that point, we might talk again about refactoring and separating those functions.
IMO i think it makes sense to keep them separately.

File: browser/js/funcs.js (hunk 1 above)
Comment (someone857) [reply browser/js/funcs.js]: done, updated to moment.js

**Comment by someone857** [comment]
@ifedapoolarewaju let me know if thats good for approval now

File: browser/js/index.js (hunk 2 above)
Comment (ifedapoolarewaju) [reply browser/js/index.js:218]: > this is not really part of the chat

But it's part of the chatList, and we are polling the chatList in order to update and render it.

> suppose in the future you want to display dates in other places and have them watched and updated automatically

I think when the need arises, we can add it for other dates. For now, I'd rather we avoid the need to add just another thing to make the codebase more complex unless we need it.

Right now, we can easily just render chatList every time we receive it (even when there's no new message), and it'll always update the time.

File: browser/js/renderers.js [hunk 4]
```diff
@@ -296,15 +296,21 @@ function renderVideoContextMenu (videoUrl) {
...
-function renderChatListItem (chatTitle, msgPreview, thumbnail, id, direction) {
+function renderChatListItem (chatTitle, msgPreview, thumbnail, id, direction, timestamp) {
   let li = document.createElement('li');
   li.classList.add('col-12', 'p-3');
 
   const msgPreviewClass = (direction == 'outward') ? 'outward' : 'inward';
 
   li.appendChild(createThumbnailDom(thumbnail));
 
-  li.appendChild(dom(`<div class="username ml-3 d-none d-sm-inline-block"><b>${chatTitle}</b><br><span class="${msgPreviewClass}">${msgPreview}</span></div>`));
```
Comment (ifedapoolarewaju) [inline browser/js/renderers.js:307]: so I mean something like this instead:

```js
li.appendChild(dom(`
  <div class="username ml-3 d-none d-sm-inline-block">
       <b>${chatTitle}</b><br>
      <span class="${msgPreviewClass}">${msgPreview}</span>
   </div>`
));
```

**Comment by someone857** [comment]
@ifedapoolarewaju done

File: browser/js/index.js [hunk 5]
```diff
@@ -61,6 +71,7 @@ document.addEventListener('DOMContentLoaded', () => {
       window.chats = chats_;
       renderChatList(window.chats);
     }
+    updateDates();
```
Comment (ifedapoolarewaju) [inline browser/js/index.js:74]: so what we could do here was to remove the `lengthMismatch` and the rest of the conditional check altogether (don't think it was really necessary anyways). So we'd always render the chatList once we have a new one. And the time will always automatically update, and we won't need the `updateDates` function anymore.

But I'll just go ahead and merge this. We can make this change subsequently. Thank you for the work!
